lib/Doctrine/Sniffs/Classes/ExceptionInterfaceNamingSniff.php

```diff
@@ -0,0 +1,110 @@
+{
+    private const CODE_NOT_AN_EXCEPTION = 'NotAnException';
+
+    /**
+     * {@inheritdoc}
+     */
+    public function register()
+    {
+        return [T_INTERFACE];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function process(File $phpcsFile, $stackPtr)
+    {
+        $importedClassNames = $this->parseImportedClassNames($phpcsFile);
+        $extendedInterfaces = $this->parseExtendedInterfaces($phpcsFile, $stackPtr);
+
+        // Set original classname instead of alias
+        $extendedInterfaces = array_map(function (string $extendedInterface) use ($importedClassNames) {
+            if (isset($importedClassNames[$extendedInterface])) {
+                return $importedClassNames[$extendedInterface];
+            }
+
+            return $extendedInterface;
+        }, $extendedInterfaces);
+
+        $hasExceptionName = strpos($phpcsFile->getDeclarationName($stackPtr), 'Exception') !== false;
+
+        $isExtendingThrowable = (in_array(Throwable::class, $importedClassNames, true) &&
+            in_array(Throwable::class, $extendedInterfaces, true)) ||
+            in_array('\\Throwable', $extendedInterfaces, true);
+
+        // Expects that an interface with the suffix "Exception" is a valid exception interface
+        $isExtendingException = count(preg_grep('/Exception$/', $extendedInterfaces)) > 0;
+
+        if ($hasExceptionName && ! $isExtendingException && ! $isExtendingThrowable) {
+            $phpcsFile->addError(
+                'Interface does not extend an exception interface',
+                $stackPtr,
+                self::CODE_NOT_AN_EXCEPTION
+            );
+
+            return;
+        }
+
+        if (! $hasExceptionName && ($isExtendingException || $isExtendingThrowable)) {
+            $phpcsFile->addError(
+                'Exception interface needs an "Exception" name suffix',
+                $stackPtr,
+                self::CODE_NOT_AN_EXCEPTION
+            );
+
+            return;
+        }
+    }
+
+    /**
+     * @return string[]
+     */
+    private function parseImportedClassNames(File $phpcsFile) : array
+    {
+        $importedClasses = [];
+        foreach (UseStatementHelper::getUseStatements($phpcsFile, 0) as $useStatement) {
+            $importedClasses[$useStatement->getNameAsReferencedInFile()] = $useStatement->getFullyQualifiedTypeName();
+        }
+
+        return $importedClasses;
+    }
+
+    /**
+     * @return string[]
+     */
+    private function parseExtendedInterfaces(File $phpcsFile, int $stackPtr) : array
+    {
+        $limit = $phpcsFile->findNext([T_OPEN_CURLY_BRACKET], $stackPtr) - 1;
+        $start = $phpcsFile->findNext([T_EXTENDS], $stackPtr, $limit);
+
+        if ($start === false) {
+            return [];
+        }
+
+        $extendedInterfaces = explode(',', $phpcsFile->getTokensAsString($start + 1, $limit - $start));
+        $extendedInterfaces = array_map('\trim', $extendedInterfaces);
```

Comment:
no leading `\` here

Reply:
I know that the coding-standard wants a dev to import functions with `use` and `trim()` is only used as a callback string here. Wouldn't removing the leading `\` lead to a lookup for `trim()` in the same namespace first?

Reply:
No, string names are always fully qualified.

https://3v4l.org/HXbil